src/Microsoft.AspNet.OData.Shared/Builder/ODataInstanceAnnotationContainer.cs

```diff
@@ -0,0 +1,131 @@
+        }
+
+        /// <summary>
+        /// Method to Add an Instance Annotation to a property
+        /// </summary>
+        /// <param name="propertyName">Name of the property</param>
+        /// <param name="annotationName">Name of Annotation</param>
+        /// <param name="value">Value of Annotation</param>
+        public void AddPropertyAnnotation(string propertyName, string annotationName, object value)
+        {
+            SetInstanceAnnotation(propertyName, annotationName, value);
+        }
+
+        /// <summary>
+        /// Get an Instance Annotation from CLR Type
+        /// </summary>
+        /// <param name="annotationName">Name of Annotation</param>
+        /// <returns>Get Annotation value for the given annotation</returns>
+        public object GetAnnotation(string annotationName)
+        {
+            return GetInstanceAnnotation(string.Empty, annotationName);
+        }
+
+        /// <summary>
+        /// Get an Instance Annotation from the Property
+        /// </summary>
+        /// <param name="propertyName">Name of the Property</param>
+        /// <param name="annotationName">Name of the Annotation</param>
+        /// <returns>Get Annotation value for the given annotation and property</returns>
+        public object GetPropertyAnnotation(string propertyName, string annotationName)
+        {
+            return GetInstanceAnnotation(propertyName, annotationName);
+        }
+
+        /// <summary>
+        /// Get All Annotations from CLR Type
+        /// </summary>
+        /// <returns>Dictionary of string(annotation name) and object value(annotation value)</returns>
+        public IDictionary<string, object> GetAllAnnotations()
+        {
+            return GetAllInstanceAnnotations(string.Empty);
+        }
+
+        /// <summary>
+        /// Get all Annotation for a Property
+        /// </summary>
+        /// <param name="propertyName">Name of Property</param>
+        /// <returns>Dictionary of string(annotation name) and object value(annotation value)</returns>
+        public IDictionary<string, object> GetAllPropertyAnnotation(string propertyName)
+        {
+            return GetAllInstanceAnnotations(propertyName);
+        }
+
+        private void SetInstanceAnnotation(string propertyName, string annotationName, object value)
+        {
+            IDictionary<string, object> annotationDictionary;
+            if (!instanceAnnotations.TryGetValue(propertyName, out annotationDictionary))
+            {
+                annotationDictionary = new Dictionary<string, object>();
+                instanceAnnotations.Add(propertyName, annotationDictionary);
+            }
+
+            if (annotationDictionary.ContainsKey(annotationName))
+            {
+                annotationDictionary[annotationName] = value;
+            }
+            else
+            {
+                annotationDictionary.Add(annotationName, value);
+            }            
```

Comment:
You should always be able to use annotationDictionary[annotationName] = value.  This will add the value if it does not already exist, otherwise overwrite the existing value.